apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloCall.java

```diff
@@ -70,6 +70,7 @@
   final List<Query> refetchQueries;
   final Optional<QueryReFetcher> queryReFetcher;
   final AtomicBoolean executed = new AtomicBoolean();
+  Optional<Callback<T>> callback = Optional.absent();
```

Comment:
What do you think to replace it with AtomicReference? And try then remove all sync blocks?